drivers: pinctrl: add bcm2711 pinctrl driver

walidbadar opened the pull request:
This pull request introduces:

* Support for the Broadcom BCM2711 pinctrl driver.
* Add PL011 UART dt nodes for BCM2711.
* Add UART pinctrl configurations for all Raspberry Pi 4B UART instances.
* Add pinctrl support for BCM2711 aux uart

<img width="561" height="220" alt="Screenshot from 2025-12-15 02-41-17" src="https://github.com/user-attachments/assets/85d371b2-a787-46b1-a801-da4ddb9ac60b" />

[input.webm](https://github.com/user-attachments/assets/dbee866d-f50a-4af7-8321-f4c92db80fde)

walidbadar committed:
drivers: pinctrl: add bcm2711 pinctrl driver

The BCM2711 GPIO controller provides 58 GPIO pins (0-57) that can be
configured for various functions including GPIO input/output and
alternate functions for peripherals like SPI, I2C, UART, PWM, etc.

Signed-off-by: Muhammad Waleed Badar <walid.badar@gmail.com>

walidbadar committed:
dts: bcm2711: add pl011 uart nodes

Add PL011 UART device tree node for BCM2711 SoC.
Node is disabled by default.

Signed-off-by: Muhammad Waleed Badar <walid.badar@gmail.com>

walidbadar committed:
boards: rpi_4b: add UART pinctrl configurations

Add pinctrl definitions for all UART interfaces on
Raspberry Pi 4B. This includes pin configurations for:
- UART0 (PL011): GPIO14/15,
- UART1 (mini UART): GPIO14/15,
- UART2-5 (PL011): Various GPIO combinations

Signed-off-by: Muhammad Waleed Badar <walid.badar@gmail.com>

JarmouniA commented on dts/bindings/pinctrl/brcm,bcm2711-pinctrl.yaml:
```diff
@@ -0,0 +1,41 @@
+
+  child-binding:
+    description: |
+      BCM2711 pin controller pin configuration group.
```
```suggestion
      Pin configuration group.
```

JarmouniA reviewed (commented):
Looks good otherwise.

walidbadar commented:
@gmarull @wearyzen could you please take a look at this PR?

walidbadar committed:
drivers: uart: add pinctrl support for bcm2711 aux uart

Add pinctrl support to the bcm2711 aux UART driver to allow
uart pin configuration via the pinctrl driver.

Signed-off-by: Muhammad Waleed Badar <walid.badar@gmail.com>

dsseng commented:
@walidbadar please follow [this guideline](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#getting-prs-reviewed), so I guess you can now reach out to Discord `#pr-help`, or first ping the assignee and some of the reviewers here.

walidbadar commented:
> @walidbadar please follow [this guideline](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#getting-prs-reviewed), so I guess you can now reach out to Discord `#pr-help`, or first ping the assignee and some of the reviewers here.

I have already pinged the reviewers/assignee here and on discord.

soburi commented on drivers/pinctrl/pinctrl_bcm2711.c:
```diff
@@ -0,0 +1,89 @@
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_MMU)) {
```
It is better to follow device_map as it handles MMU disabled abstractly.

walidbadar commented on boards/raspberrypi/rpi_4b/rpi_4b.dts:
```diff
@@ -40,6 +43,15 @@
 	status = "okay";
+};
+
+&uart2 {
```
I have removed `uart2` node.
